사용자는 노티를 볼 수 있다.

stompesi opened the pull request:
#1
- 사용자는 자신이 행동 한것 이외에 참여한 공구에 대하여 알림을 볼 수있습니다. (공구참여, 공구마감)
- 사용자는 해당 알람을 누를 시 해당알람은 "읽음" 으로 변합니다. (Text-Bold 해제이용)
- 사용자는 해당 알림을 누를 시 해당게시글로 이동합니다. (스크롤이동)
- 사용자는 최근 8개의 알람까지 볼 수 있습니다.

Commit:
* 사용자는 노티를 볼 수 있다.
 - 사용자는 자신이 행동 한것 이외에 참여한 공구에 대하여 알림을 볼 수있습니다. (공구참여, 공구마감)
 - 사용자는 해당 알람을 누를 시 해당알람은 "읽음" 으로 변합니다. (Text-Bold 해제)
 - 사용자는 해당 알림을 누를 시 해당게시글로 이동합니다. (스크롤이동)
 - 사용자는 최근 8개의 알람까지 볼 수 있습니다.

minhyeok4dev commented:
1. Notification 이라는 명칭이 종모양 아이콘과 따로 구성되어있어 이질감을 줍니다. Notification 텍스트를 삭제하거나 종 모양 li 태그와 합쳐야할것 같습니다.
   ![screen shot 2015-02-03 at 7 20 09 pm](https://cloud.githubusercontent.com/assets/3871307/6032845/248a1c12-abdc-11e4-9123-32b2819ccb62.png)

2.마감하기, 마감풀기 때 노티 0이 표시되어 나타납니다.
![screen shot 2015-02-03 at 7 20 25 pm](https://cloud.githubusercontent.com/assets/3871307/6032849/33c3daa6-abdc-11e4-82da-38d65dd976c2.png)

이게바로 저번에 멘토님께서 보여주고자 하셨던 자바스크립트 이슈 때문인데요.

``` ruby
$(document).ready(function(){...
```

로 시작하는 것은 레일즈 터보링크 문제가 있어

``` ruby
$(document).on('ready page:load', function(){...
```

로 바꿔줘야 합니다. 콜백 바인딩은 on으로 설정하는것이 권장되는데 HTML DOM 렌더링 관련해서 나름 중요한 이슈였던걸로 기억이 나네요. 한번 찾아보셔도 좋을것같습니다

3.notification routes가 중복됩니다. 지금 현재 이렇게 정의되어있죠

``` ruby
resources :posts do
    member do
      get 'close'
      get 'open'
    end
    resources :comments, only: [:create]
    resources :notifications, only: [:create]
  end

  resources :notifications, only: [:update]
```

쓸데없는 라우트를 뺴려고 똑같은 notification을 action만 다르게 중복으로 정의했는데, 이는 사실 shallow를 이용하여 

``` ruby
resources :posts do
    member do
      get 'close'
      get 'open'
    end
    resources :comments, only: [:create]
    resources :notifications, only: [:create, :update], shallow: true
  end

  # resources :notifications, only: [:update]
```

로 똑같이 표현할 수 있습니다. 이로인해 라우트 문제를 똑같이 해결할 수 있습니다.

위에선 shallow를 notification에만 걸었는데 상위 posts에 대해 그 안에 있는 모든 namespace에 대해서도 shallow를 적용할 수 있습니다. 개인적으로는 상위에 걸어주는게 더 나을것같습니다.

angdev commented:
:+1:

stompesi commented:
1. Notification명칭을 빼는것을 실수로 빼지 않았네요 삭제하겠습니다. 
2. '레일즈 터보링크 문제' 라는것은 따로 한번 찾아보겠습니다.
3. 마찬가지로 shallow에 대한 개념도 찾아봐야겠습니다.

Commit:
코멘트 수정사항 수정 & 추가 오류 수정
- /admin/users 에서 notifications 때문에 오류발생하는 문제 해결
- notification과 종 icon 중복 제거
- routes 중복 제거
- javascripts 레일즈 터보링크 해결

shaynekang commented:
수고하셨습니다~
1. 자바스크립트에서는 `click`보다 `on('click')`을 권장합니다. 후자는 element가 사라졌다 다시 생겨도 정상적으로 동작합니다. [notification.js](https://github.com/HwaYo/soma09/blob/8c5255004c506345743ae5b639e0606ae9881806/app/assets/javascripts/notifications.js)의 관련 코드를 수정하면 좋을 것 같습니다.
2. 다소 견해가 엇갈립니디만, 자바스크립트의 컨벤션은 보통 카멜케이스(camelCase)입니다. [notification.js](https://github.com/HwaYo/soma09/blob/8c5255004c506345743ae5b639e0606ae9881806/app/assets/javascripts/notifications.js)에서 `new_noti_number`보다는 `newNotiNumber`가 좋을 것 같습니다.
3. 반면 CSS의 컨벤션은 소문자+하이픈입니다. `new_noti_number`보다는 `new-noti-number`가 좋을 것 같습니다.
4. 개인적으로는 축약어를 좋아하지 않는데, 다른 사람들이 코드를 리뷰할 때 어려워지기 때문입니다. noti는 notification으로 풀어쓰면 좋을 것 같네요.
5. 컨트롤러끼리 서로를 호출하는 건 권장하지 않습니다. 코드의 설계가 꼬이는 문제가 생깁니다. [NotificationsController](https://github.com/HwaYo/soma09/blob/8c5255004c506345743ae5b639e0606ae9881806/app/controllers/notifications_controller.rb)의 `send_notification`은 `Post`모델로 옮기면 좋겠네요.(차후에는 [서비스 오브젝트](http://blog.codeclimate.com/blog/2012/10/17/7-ways-to-decompose-fat-activerecord-models/)를 이용하게 될겁니다.)
6. [NotificationsController](https://github.com/HwaYo/soma09/blob/8c5255004c506345743ae5b639e0606ae9881806/app/controllers/notifications_controller.rb)의 `send_notification`에서 `@notification.save`가 실패했을 경우 false만 반환하고 에러를 씹어버리게 됩니다. 이러먼 이 코드에서 문제가 생겼을 경우 버그를 찾기 어려워지죠. 이런 경우는 예외처리를 해주거나, 뱅(!) 메소드를 사용해서 강제로 예외를 터뜨려주는 것이 좋습니다. (ex: `@notification.save!`)
7. 뷰에서 루비 코드가 많아지면 차후에 유지보수가 어려워집니다. 뷰에서 `current_user.notifications`에 관한 코드는 컨트롤러로 옮길 수 있으면 좋겠네요. ㅎㅎ
8. 컨트롤러에서 `current_user`를 사용할 땐 굳이 `@user`로 받아올 필요는 없을 듯 합니다. 그대로 `current_user`를 사용하셔도 됩니다. ㅎㅎ
9. 노티피케이션을 랜더링할 때 `unless current_user == nil`이 있는데 current_user가 nil인 경우가 있나요? 있으면 다른 의미로 문제일 것 같은데... ㅡ,.ㅡ

shaynekang commented:
1. [notifications.js](https://github.com/HwaYo/soma09/blob/8c5255004c506345743ae5b639e0606ae9881806/app/assets/javascripts/notifications.js)에서 다음의 코드가 중복되는 것 같네요. 리펙토링을 해주면 좋을 것 같습니다.
   
   ``` javascript
   new_noti_number = $("#new_noti_number").text();
   if(new_noti_number == 0) {
     $("#new_noti_number").remove();
   }
   
   // ...
   
   new_noti_number = $("#new_noti_number").text() - 1;
   if(new_noti_number === 0) {
     $("#new_noti_number").remove();
   } else {
     $("#new_noti_number").text(new_noti_number + "");
   }
   ```
   
   그리고 지금은 `$("#new_noti_number")`를 여러번 부르고 있는데, 한 번만 불러주는 것이 성능상 더 좋습니다. 참고해주세요. ㅎㅎ
2. `NotificationsController`의 `send_notification`은 다음과 같이 고칠 수 있습니다.
   
   ``` ruby
   def self.send_notification(post, send_user, message)
     post.participants.each do |participant|
       next unless participant.user == send_user
   
       Notification.create!({
         message: message,
         target_user: participant.user,
         send_user: send_user,
         post: post
       })
     end
   end
   ```
3. `posts#create`에서 `@new_post.save`와 `@new_post.participants.find_or_create_by(user: @user)`를 동시에 호출하면 사실상 쿼리가 두 번 날라가게 됩니다. 이런 경우는 transaction을 걸어줘야 하는데,
   
   ``` ruby
   Post.transaction do
     if @new_post.save
       @new_post.participants.find_or_create_by(user: @user)
       redirect_to posts_path
     else
       @posts = Post.latest
       render 'index'
     end
   end
   ```
   
   이렇게 Model.transaction으로 감싸주는 방법이 있는데, 코드가 덜 깔끔한 편입니다.
   이런 경우 레일즈 association을 잘 활용하면 transaction이 자동으로 걸리게 할 수 있습니다.
   
   ``` ruby
   @new_post = Post.new(post_params)
   
   @new_post.user = current_user
   @new_post.participants.build(user: current_user)
   
   if @new_post.save
     redirect_to posts_path
   else
     @posts = Post.latest
     render 'index'
   end
   ```
   
   이렇게 association의 `build` 메소드를 활용하면 1. 부모의 foreign_key를 자동으로 대입해주고, 2. 부모를 insert/update할 경우 자식도 자동으로 insert/update 해줍니다. 겸사겸사 transaction도 걸어주죠.
   코드가 더 깔끔하기 때문에, 이렇게 연관 모델들을 입력/갱신하는 걸 권장합니다.

이상입니다. 파팅파팅 하세요~!

Commit:
- 리팩토링
1. application.html.erb 에 있던  notification 코드 partitial로 변경 및 View에서 작업하던 ruby code 컨트롤러로 이동
2. notifications.js -> on('click')으로 코드수정, 네이밍 변경
3. current_user 사용

stompesi commented:
아 많이 못짯네요..ㅠ 다음엔 좀더 신경 쓰겠습니다!
그런데 궁금사항이 있어 질문드립니다. 
1. 컨트롤러끼리 서로를 호출하는 건 권장하지 않습니다. 코드의 설계가 꼬이는 문제가 생깁니다. NotificationsController의 send_notification은 Post모델로 옮기면 좋겠네요.(차후에는 서비스 오브젝트를 이용하게 될겁니다.)
   -> 
   1. participants_controller.rb
      message = "#{current_user.name}님이 공동구매에 참여하였습니다."
      NotificationsController.send_notification(@post, current_user, message)
   2. posts_controller.rb
      message = "참가하셨던 공동구매가 마감되었습니다!"
      NotificationsController.send_notification(@post, @post.user, message)
   
   이런식으로 2개가 있어 코드 중복을 막고싶어서 사용했는데 컨트롤러호출없이 사용하는 방법은 없을까요?

Commit:
모호한 표현 제거

shaynekang commented:
음 예를 들면 이런거죠. ㅎㅎ

``` ruby
# models/participant.rb
class Participant < ActiveRecord::Base
  # ...

  def self.send_notification(send_user, message)
    self.all.each do |participant|
      next if participant.user == send_user

      Notification.create!({
        message: message,
        target_user: participant.user,
        send_user: send_user,
        post: post
      })
    end
  end
end
```

``` ruby
class PostsController < ApplicationController
  # ...

  def close
    @post = Post.find(params[:id])

    if @post.update(closed: true)
      # ...

      message = "참가하셨던 공동구매가 마감되었습니다!"
      @post.participants.send_notification(@post.user, message)
    end

    redirect_to posts_path
  end

  # ...
end

```

이렇게 participant.rb 에 클래스 메소드로 정의해 놓은 후, `@post.participants.send_notification`을 하면 호출이 됩니다. ㅎㅎ

stompesi committed:
컨트롤러끼리 호출 제거, 오타수정
